modules/database/funcs.go

```diff
@@ -99,11 +101,30 @@ func (d *Database) GetDatabase() *mariadbv1.MariaDBDatabase {
 
 //
 // CreateOrPatchDB - create or patch the service DB instance
+// Deprecated. Use CreateOrPatchDBWithLabel instead and specify empty labels
+// if you want to create the DB in the default DB service.
 //
 func (d *Database) CreateOrPatchDB(
 	ctx context.Context,
 	h *helper.Helper,
 ) (ctrl.Result, error) {
+	return d.CreateOrPatchDBWithLabel(ctx, h, map[string]string{})
+}
+
+//
+// CreateOrPatchDBWithLabel - create or patch the service DB instance on
+// the DB service selected via labels. If the labels is empty then the
+// default DB service (labelled with cr=mariadb-openstack) will be used.
+//
+func (d *Database) CreateOrPatchDBWithLabel(
+	ctx context.Context,
+	h *helper.Helper,
+	labels map[string]string,
+) (ctrl.Result, error) {
+
+	if len(labels) == 0 {
+		labels["cr"] = "mariadb-openstack"
```

Comment:
so the useage for nova_cell1 or nova_cell2 will be to set 

labels["app"] = "mariadb"
lables["cr"] = "mariadb-nova-cell-1" 

that will be used as a selector as if we had done 

selector := map[string]string{
		"app": "mariadb",
        "cr": "mariadb-nova-cell-1" 
	}

when invoking service.GetServicesListWithLabel on line 66.

although we actually wont need to manually set app since you have set it on line 60.

Reply:
I did not want to change the "app" part of the selector to keep the existing behavior, i.e. not to break the existing operators. But in general we might want to consider using a more generic lookup to support both the adoption and the external DB use cases. As the interface now takes a label selector map it is easy to change lib common to use different default lookups if needed.

Yes, the NovaCell CR will have a "DB label selector" field that can be set like "cr":"mariadb-nova-cell-1" during deployment. And then the controller would use that info to call CreateOrPatchDBWithLabel and select the specific DB Service instance to use for the given nova cell instance.

Reply:
I should have said, the app=mariadb in CreateOrPatchDB wouldn't be default, that would simply be hardcoded behavior as it is right now. But moving this hardcoding from setDatabaseHostname to CreateOrPatchDB would mean that CreateOrPatchDBWithLabel would be unaffected by that hardcoding.

Reply:
And CreateOrPatchDBWithLabel when `len(labels) == 0` should perhaps just fail.

Reply:
> Re defaults, how about putting app=mariadb default to CreateOrPatchDB - that way we preserve current behavior for users of that method. And then not handling anything for len(labels) == 0 in CreateOrPatchDBWithLabel. That means anyone who is switching from CreateOrPatchDB to CreateOrPatchDBWithLabel will have to provide all their desired labels explicitly.

Make sense I will do it.

> Another topic: From the adoption perspective, i'm wondering how could we support that use case -- prioritizing external adoption DB connection, and only if that doesn't exist, using podified MariaDB connection. Could we be doing two lookups here, first by adding some magic label "adoption=yes" (or something like that slightly_smiling_face ) to the set of provided labels, and only if that doesn't exist, look up the service by the provided labels only?

Nova will need to expose the label selector for DB Service instances on the CR as Nova needs to support multiple DB Service instances. So for Nova using a DB would mean setting the specific labels in the CR that matches with the Service instance that points to DB service to be used. So at start the labels in the Nova CR could be configured to select the Service instance that points to the external DB. And when we switch to the internal one the label in the Nova CR need to be changed to match with the internal Service instance.

For those service operators where a single (default) DB Service instance is used and therefore now label selector is provided on the CR we can do what you suggest and implement a logic and look for two set of default labels in precedence order. But alternatively we can keep a single default labels in this query and instead change the label on the Service instance to match the default query. So at deployment with external DB the a Service instance mapped to the external DB is created with the label that will match our default query. So the service operators will use that DB by default. Then when we deploy the internal DB then a new Service instance is created for that DB first without matching labels. When we switch we from external to internal, we need to change the labels on the external Service instance not to match the default query any more and change the label on the internal Service instance to match the default query. Then we probably need to kick the service operators to detect the change and reconcile to the new DB url.

Reply:
> But alternatively we can keep a single default labels in this query and instead change the label on the Service instance to match the default query. So at deployment with external DB the a Service instance mapped to the external DB is created with the label that will match our default query. So the service operators will use that DB by default. Then when we deploy the internal DB then a new Service instance is created for that DB first without matching labels. When we switch we from external to internal, we need to change the labels on the external Service instance not to match the default query any more and change the label on the internal Service instance to match the default query. Then we probably need to kick the service operators to detect the change and reconcile to the new DB url.

+1 that sounds good to me.

Reply:
Moved the defaulting.